src/sqlancer/cnosdb/CnosDBComparatorHelper.java

```diff
@@ -0,0 +1,149 @@
+import java.util.List;
+import java.util.Set;
+import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
+
+import sqlancer.IgnoreMeException;
+import sqlancer.Main;
+import sqlancer.cnosdb.client.CnosDBResultSet;
+import sqlancer.cnosdb.query.CnosDBSelectQuery;
+import sqlancer.common.query.ExpectedErrors;
+
+public final class CnosDBComparatorHelper {
+
+    private CnosDBComparatorHelper() {
+    }
+
+    public static List<String> getResultSetFirstColumnAsString(String queryString, ExpectedErrors errors,
+            CnosDBGlobalState state) throws Exception {
+        if (state.getOptions().logEachSelect()) {
+            // TODO: refactor me
+            state.getLogger().writeCurrent(queryString);
+            try {
+                state.getLogger().getCurrentFileWriter().flush();
+            } catch (IOException e) {
+                // TODO Auto-generated catch block
+                e.printStackTrace();
+            }
+        }
+        CnosDBSelectQuery q = new CnosDBSelectQuery(queryString, errors);
+        List<String> result = new ArrayList<>();
+        CnosDBResultSet resultSet;
+        try {
+            q.executeAndGet(state);
+            resultSet = q.getResultSet();
+            if (resultSet == null) {
+                throw new AssertionError(q);
+            }
+            while (resultSet.next()) {
+                result.add(resultSet.getString(1));
+            }
+        } catch (Exception e) {
+            if (e instanceof IgnoreMeException) {
+                throw e;
+            }
+            if (e instanceof NumberFormatException) {
+                throw new IgnoreMeException();
+            }
+            if (e.getMessage() == null) {
+                throw new AssertionError(queryString, e);
+            }
+            if (errors.errorIsExpected(e.getMessage())) {
+                throw new IgnoreMeException();
+            }
+            throw new AssertionError(queryString, e);
+        }
+
+        return result;
+    }
+
+    public static void assumeResultSetsAreEqual(List<String> resultSet, List<String> secondResultSet,
+            String originalQueryString, List<String> combinedString, CnosDBGlobalState state) {
+        if (resultSet.size() != secondResultSet.size()) {
+            String queryFormatString = "-- %s;\n-- cardinality: %d";
+            String firstQueryString = String.format(queryFormatString, originalQueryString, resultSet.size());
+            String secondQueryString = String.format(queryFormatString, String.join(";", combinedString),
+                    secondResultSet.size());
+            state.getState().getLocalState().log(String.format("%s\n%s", firstQueryString, secondQueryString));
+            String assertionMessage = String.format("the size of the result sets mismatch (%d and %d)!\n%s\n%s",
+                    resultSet.size(), secondResultSet.size(), firstQueryString, secondQueryString);
+            Main.nrUnsuccessfulActions.addAndGet(1);
```

Comment:
The most recent commit added a couple of `Main.nrUnsuccessfulActions.addAndGet(1);` and similar calls like this. I believe we should remove them to not let the two comparator helper classes diverge until we fix the code duplication problem.